pennylane_lightning/lightning_qubit/_measurements.py

```diff
@@ -251,7 +268,168 @@ def measure_final_state(self, circuit: QuantumScript) -> Result:
+        self,
+        mps: List[SampleMeasurement],
+        shots: Shots,
+    ) -> TensorLike:
+        """
+        Returns the samples of the measurement process performed on the given state,
+        by rotating the state into the measurement basis using the diagonalizing gates
+        given by the measurement process.
+
+        Args:
+            mps (~.measurements.SampleMeasurement): The sample measurements to perform
+            shots (~.measurements.Shots): The number of samples to take
+
+        Returns:
+            TensorLike[Any]: Sample measurement results
+        """
+        # apply diagonalizing gates
+        self._apply_diagonalizing_gates(mps)
+
+        total_indices = self._qubit_state.num_wires
+        wires = qml.wires.Wires(range(total_indices))
+
+        def _process_single_shot(samples):
+            processed = []
+            for mp in mps:
+                res = mp.process_samples(samples, wires)
+                if not isinstance(mp, CountsMP):
+                    res = qml.math.squeeze(res)
+
+                processed.append(res)
+
+            return tuple(processed)
+
+        # if there is a shot vector, build a list containing results for each shot entry
+        if shots.has_partitioned_shots:
+            processed_samples = []
+            for s in shots:
+                # currently we call sample_state for each shot entry, but it may be
+                # better to call sample_state just once with total_shots, then use
+                # the shot_range keyword argument
+                try:
+                    if self._mcmc:
+                        samples = self._measurement_lightning.generate_mcmc_samples(
+                            len(wires), self._kernel_name, self._num_burnin, s
+                        ).astype(int, copy=False)
+                    else:
+                        samples = self._measurement_lightning.generate_samples(
+                            len(wires), s
+                        ).astype(int, copy=False)
+                except ValueError as e:
+                    if str(e) != "probabilities contain NaN":
+                        raise e
+                    samples = qml.math.full((s, len(wires)), 0)
+
+                processed_samples.append(_process_single_shot(samples))
+            self._apply_diagonalizing_gates(mps, adjoint=True)
+            return tuple(zip(*processed_samples))
+
+        try:
+            if self._mcmc:
+                samples = self._measurement_lightning.generate_mcmc_samples(
+                    len(wires), self._kernel_name, self._num_burnin, shots.total_shots
+                ).astype(int, copy=False)
+            else:
+                samples = self._measurement_lightning.generate_samples(
+                    len(wires), shots.total_shots
+                ).astype(int, copy=False)
+        except ValueError as e:
+            if str(e) != "probabilities contain NaN":
+                raise e
```

Comment:
I believe this is only there to catch 0-prob cases arising from post-selected MCMs. We do not raise because we want the computation to continue and preserve the auto-differentiability of the `defer_measurement` transform.

Reply:
I should say, we do not test or officially support the transform yet, but this is a short-term goal so I would keep this there by analogy with DQ.